Failures in test_retry_in_subprocess

There is random (yet regular) errors with `test_retry_in_subprocess` in CI environment (see e.g., https://gitlab.esrf.fr/silx/bob/silx/-/jobs/193581):
```
ERROR: test_retry_in_subprocess (silx.io.test.test_h5py_utils.TestH5pyUtils) (mode='w', libver='latest')
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/builds/silx/bob/silx/silx-0.15.0rc1/build/debian11/silx-0.15.0rc1/.pybuild/cpython3_3.8_silx/build/silx/utils/retry.py", line 63, in _handle_exception
    yield
  File "/builds/silx/bob/silx/silx-0.15.0rc1/build/debian11/silx-0.15.0rc1/.pybuild/cpython3_3.8_silx/build/silx/utils/retry.py", line 242, in wrapper
    raise result
silx.utils.retry.RetryError
The above exception was the direct cause of the following exception:
Traceback (most recent call last):
  File "/builds/silx/bob/silx/silx-0.15.0rc1/build/debian11/silx-0.15.0rc1/.pybuild/cpython3_3.8_silx/build/silx/io/test/test_h5py_utils.py", line 120, in wrapper
    test(self)
  File "/builds/silx/bob/silx/silx-0.15.0rc1/build/debian11/silx-0.15.0rc1/.pybuild/cpython3_3.8_silx/build/silx/io/test/test_h5py_utils.py", line 379, in test_retry_in_subprocess
    names = top_level_names_test(txtfilename, filename, **kw)
  File "/builds/silx/bob/silx/silx-0.15.0rc1/build/debian11/silx-0.15.0rc1/.pybuild/cpython3_3.8_silx/build/silx/utils/retry.py", line 227, in wrapper
    for options in _retry_loop(
  File "/builds/silx/bob/silx/silx-0.15.0rc1/build/debian11/silx-0.15.0rc1/.pybuild/cpython3_3.8_silx/build/silx/utils/retry.py", line 90, in _retry_loop
    raise RetryTimeoutError from options.get("exception")
silx.utils.retry.RetryTimeoutError
``

Also got on power9:
```
FAIL: test_retry_in_subprocess (silx.io.test.test_h5py_utils.TestH5pyUtils) (mode='w')
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/home/tvincent/silx/build/lib.linux-ppc64le-3.9/silx/io/test/test_h5py_utils.py", line 120, in wrapper
    test(self)
  File "/home/tvincent/silx/build/lib.linux-ppc64le-3.9/silx/io/test/test_h5py_utils.py", line 391, in test_retry_in_subprocess
    top_level_names_test(txtfilename, filename, **kw)
AssertionError: RetryTimeoutError not raised
```

--- a/silx/io/test/test_h5py_utils.py
+++ b/silx/io/test/test_h5py_utils.py
@@ -31,6 +31,7 @@ __date__ = "27/01/2020"
 import unittest
 import os
 import sys
+import time
 import shutil
 import tempfile
 import threading
@@ -96,18 +97,20 @@ def _top_level_names_test(txtfilename, *args, **kw):
     sys.stderr = open(os.devnull, "w")
 
     with open(txtfilename, mode="r") as f:
-        counter = int(f.readline().strip())
+        failcounter = int(f.readline().strip())
 
-    nsleep = kw.pop("nsleep")
-    if counter < nsleep:
-        counter += 1
+    ncausefailure = kw.pop("ncausefailure")
+    faildelay = kw.pop("faildelay")
+    if failcounter < ncausefailure:
+        time.sleep(faildelay)
+        failcounter += 1
         with open(txtfilename, mode="w") as f:
-            f.write(str(counter))
-        if counter % 2:
+            f.write(str(failcounter))
+        if failcounter % 2:
             raise RetryError
         else:
             _cause_segfault()
-    return h5py_utils.top_level_names(*args, **kw)
+    return h5py_utils._top_level_names(*args, **kw)
 
 
 top_level_names_test = h5py_utils.retry_in_subprocess()(_top_level_names_test)
@@ -145,7 +148,7 @@ class TestH5pyUtils(unittest.TestCase):
         i = 1
         while True:
             filename = os.path.join(self.test_dir, "file{}.h5".format(i))
-            with self._open_context(filename) as f:
+            with self._open_context(filename):
                 pass
             yield filename
             i += 1
@@ -326,34 +329,28 @@ class TestH5pyUtils(unittest.TestCase):
     @subtests
     def test_retry_custom(self):
         filename = self._new_filename()
-        nsleep = 3
-        retry_period = 0.01
-
-        # just to make sure the test doesn't hang
-        overhead = 10
+        ncausefailure = 3
+        faildelay = 0.1
+        sufficient_timeout = ncausefailure * (faildelay + 10)
+        insufficient_timeout = ncausefailure * faildelay * 0.5
 
         @h5py_utils.retry_contextmanager()
         def open_item(filename, name):
-            nonlocal counter
-            if counter < nsleep:
-                counter += 1
+            nonlocal failcounter
+            if failcounter < ncausefailure:
+                time.sleep(faildelay)
+                failcounter += 1
                 raise RetryError
             with h5py_utils.File(filename) as h5file:
                 yield h5file[name]
 
-        counter = 0
-        kw = {
-            "retry_timeout": nsleep * (retry_period + overhead),
-            "retry_period": retry_period,
-        }
+        failcounter = 0
+        kw = {"retry_timeout": sufficient_timeout}
         with open_item(filename, "/check", **kw) as item:
             self.assertTrue(item[()])
 
-        counter = 0
-        kw = {
-            "retry_timeout": nsleep * (retry_period * 0.9),
-            "retry_period": retry_period,
-        }
+        failcounter = 0
+        kw = {"retry_timeout": insufficient_timeout}
         with self.assertRaises(RetryTimeoutError):
             with open_item(filename, "/check", **kw) as item:
                 pass
@@ -361,18 +358,17 @@ class TestH5pyUtils(unittest.TestCase):
     @subtests
     def test_retry_in_subprocess(self):
         filename = self._new_filename()
-        txtfilename = os.path.join(self.test_dir, "counter.txt")
-        nsleep = 3
-        retry_period = 0.01
-
-        # just to make sure the test doesn't hang
-        overhead = 10
+        txtfilename = os.path.join(self.test_dir, "failcounter.txt")
+        ncausefailure = 3
+        faildelay = 0.1
+        sufficient_timeout = ncausefailure * (faildelay + 10)
+        insufficient_timeout = ncausefailure * faildelay * 0.5
 
         kw = {
-            "retry_timeout": nsleep * (retry_period + overhead),
-            "retry_period": retry_period,
+            "retry_timeout": sufficient_timeout,
             "include_only": None,
-            "nsleep": nsleep,
+            "ncausefailure": ncausefailure,
+            "faildelay": faildelay,
         }
         with open(txtfilename, mode="w") as f:
             f.write("0")
@@ -380,10 +376,10 @@ class TestH5pyUtils(unittest.TestCase):
         self.assertEqual(names, ["check"])
 
         kw = {
-            "retry_timeout": nsleep * (retry_period * 0.9),
-            "retry_period": retry_period,
+            "retry_timeout": insufficient_timeout,
             "include_only": None,
-            "nsleep": nsleep,
+            "ncausefailure": ncausefailure,
+            "faildelay": faildelay,
         }
         with open(txtfilename, mode="w") as f:
             f.write("0")

--- a/silx/utils/retry.py
+++ b/silx/utils/retry.py
@@ -189,7 +189,8 @@ def retry_in_subprocess(
 ):
     """Same as `retry` but it also retries segmentation faults.
 
-    On Window you cannot use this decorator with the "@" syntax:
+    As subprocesses are spawned, you cannot use this decorator with the "@" syntax
+    because the decorated method needs to be an attribute of a module:
 
     .. code-block:: python
 
@@ -214,14 +215,16 @@ def retry_in_subprocess(
             _retry_period = kw.pop("retry_period", retry_period)
             _retry_on_error = kw.pop("retry_on_error", retry_on_error)
 
-            queue = multiprocessing.Queue(maxsize=1)
+            ctx = multiprocessing.get_context('spawn')
+
+            queue = ctx.Queue(maxsize=1)
             prockw = {
                 "target": _subprocess_main,
                 "args": (queue, method, retry_on_error) + args,
                 "kwargs": kw,
             }
 
-            p = multiprocessing.Process(**prockw)
+            p = ctx.Process(**prockw)
             p.start()
             try:
                 for options in _retry_loop(
@@ -229,7 +232,7 @@ def retry_in_subprocess(
                 ):
                     with _handle_exception(options):
                         if not p.is_alive():
-                            p = multiprocessing.Process(**prockw)
+                            p = ctx.Process(**prockw)
                             p.start()
                         try:
                             result = queue.get(block=True, timeout=_retry_period)

--- a/silx/utils/retry.py
+++ b/silx/utils/retry.py
@@ -90,7 +90,9 @@ def _retry_loop(retry_timeout=None, retry_period=None, retry_on_error=None):
             raise RetryTimeoutError from options.get("exception")
 
 
-def retry(retry_timeout=None, retry_period=None, retry_on_error=_default_retry_on_error):
+def retry(
+    retry_timeout=None, retry_period=None, retry_on_error=_default_retry_on_error
+):
     """Decorator for a method that needs to be executed until it not longer
     fails or until `retry_on_error` returns False.
 
@@ -215,25 +217,33 @@ def retry_in_subprocess(
             _retry_period = kw.pop("retry_period", retry_period)
             _retry_on_error = kw.pop("retry_on_error", retry_on_error)
 
-            ctx = multiprocessing.get_context('spawn')
+            ctx = multiprocessing.get_context("spawn")
 
-            queue = ctx.Queue(maxsize=1)
-            prockw = {
-                "target": _subprocess_main,
-                "args": (queue, method, retry_on_error) + args,
-                "kwargs": kw,
-            }
+            def start_subprocess():
+                queue = ctx.Queue(maxsize=1)
+                p = ctx.Process(
+                    target=_subprocess_main,
+                    args=(queue, method, retry_on_error) + args,
+                    kwargs=kw,
+                )
+                p.start()
+                return p, queue
 
-            p = ctx.Process(**prockw)
-            p.start()
+            def stop_subprocess(p):
+                try:
+                    p.kill()
+                except AttributeError:
+                    p.terminate()
+                p.join()
+
+            p, queue = start_subprocess()
             try:
                 for options in _retry_loop(
                     retry_timeout=_retry_timeout, retry_on_error=_retry_on_error
                 ):
                     with _handle_exception(options):
                         if not p.is_alive():
-                            p = ctx.Process(**prockw)
-                            p.start()
+                            p, queue = start_subprocess()
                         try:
                             result = queue.get(block=True, timeout=_retry_period)
                         except Empty:
@@ -242,15 +252,12 @@ def retry_in_subprocess(
                             pass
                         else:
                             if isinstance(result, BaseException):
+                                stop_subprocess(p)
                                 raise result
                             else:
                                 return result
             finally:
-                try:
-                    p.kill()
-                except AttributeError:
-                    p.terminate()
-                p.join()
+                stop_subprocess(p)
 
         return wrapper
 

--- a/silx/utils/test/test_retry.py
+++ b/silx/utils/test/test_retry.py
@@ -31,6 +31,7 @@ __date__ = "05/02/2020"
 import unittest
 import os
 import sys
+import time
 import tempfile
 
 from .. import retry
@@ -47,19 +48,20 @@ def _cause_segfault():
         c += 1
 
 
-def _submain(filename, kwcheck=None, nsleep=0):
+def _submain(filename, kwcheck=None, ncausefailure=0, faildelay=0):
     assert filename
     assert kwcheck
     sys.stderr = open(os.devnull, "w")
 
     with open(filename, mode="r") as f:
-        counter = int(f.readline().strip())
+        failcounter = int(f.readline().strip())
 
-    if counter < nsleep:
-        counter += 1
+    if failcounter < ncausefailure:
+        time.sleep(faildelay)
+        failcounter += 1
         with open(filename, mode="w") as f:
-            f.write(str(counter))
-        if counter % 2:
+            f.write(str(failcounter))
+        if failcounter % 2:
             raise retry.RetryError
         else:
             _cause_segfault()
@@ -72,7 +74,7 @@ _wsubmain = retry.retry_in_subprocess()(_submain)
 class TestRetry(unittest.TestCase):
     def setUp(self):
         self.test_dir = tempfile.mkdtemp()
-        self.ctr_file = os.path.join(self.test_dir, "counter.txt")
+        self.ctr_file = os.path.join(self.test_dir, "failcounter.txt")
 
     def tearDown(self):
         if os.path.exists(self.ctr_file):
@@ -80,97 +82,86 @@ class TestRetry(unittest.TestCase):
         os.rmdir(self.test_dir)
 
     def test_retry(self):
-        nsleep = 3
-        retry_period = 0.01
-
-        # just to make sure the test doesn't hang
-        overhead = 10
+        ncausefailure = 3
+        faildelay = 0.1
+        sufficient_timeout = ncausefailure * (faildelay + 10)
+        insufficient_timeout = ncausefailure * faildelay * 0.5
 
         @retry.retry()
         def method(check, kwcheck=None):
             assert check
             assert kwcheck
-            nonlocal counter
-            if counter < nsleep:
-                counter += 1
+            nonlocal failcounter
+            if failcounter < ncausefailure:
+                time.sleep(faildelay)
+                failcounter += 1
                 raise retry.RetryError
             return True
 
-        counter = 0
+        failcounter = 0
         kw = {
             "kwcheck": True,
-            "retry_timeout": nsleep * (retry_period + overhead),
-            "retry_period": retry_period,
+            "retry_timeout": sufficient_timeout,
         }
         self.assertTrue(method(True, **kw))
 
-        counter = 0
+        failcounter = 0
         kw = {
             "kwcheck": True,
-            "retry_timeout": nsleep * (retry_period * 0.9),
-            "retry_period": retry_period,
+            "retry_timeout": insufficient_timeout,
         }
         with self.assertRaises(retry.RetryTimeoutError):
             method(True, **kw)
 
     def test_retry_contextmanager(self):
-        nsleep = 3
-        retry_period = 0.01
-
-        # just to make sure the test doesn't hang
-        overhead = 10
+        ncausefailure = 3
+        faildelay = 0.1
+        sufficient_timeout = ncausefailure * (faildelay + 10)
+        insufficient_timeout = ncausefailure * faildelay * 0.5
 
         @retry.retry_contextmanager()
         def context(check, kwcheck=None):
             assert check
             assert kwcheck
-            nonlocal counter
-            if counter < nsleep:
-                counter += 1
+            nonlocal failcounter
+            if failcounter < ncausefailure:
+                time.sleep(faildelay)
+                failcounter += 1
                 raise retry.RetryError
             yield True
 
-        counter = 0
-        kw = {
-            "kwcheck": True,
-            "retry_timeout": nsleep * (retry_period + overhead),
-            "retry_period": retry_period,
-        }
+        failcounter = 0
+        kw = {"kwcheck": True, "retry_timeout": sufficient_timeout}
         with context(True, **kw) as result:
             self.assertTrue(result)
 
-        counter = 0
-        kw = {
-            "kwcheck": True,
-            "retry_timeout": nsleep * (retry_period * 0.9),
-            "retry_period": retry_period,
-        }
+        failcounter = 0
+        kw = {"kwcheck": True, "retry_timeout": insufficient_timeout}
         with self.assertRaises(retry.RetryTimeoutError):
             with context(True, **kw) as result:
                 pass
 
     def test_retry_in_subprocess(self):
-        nsleep = 3
-        retry_period = 0.01
-
-        # just to make sure the test doesn't hang
-        overhead = 10
+        ncausefailure = 3
+        faildelay = 0.1
+        sufficient_timeout = ncausefailure * (faildelay + 10)
+        insufficient_timeout = ncausefailure * faildelay * 0.5
 
         kw = {
-            "nsleep": nsleep,
+            "ncausefailure": ncausefailure,
+            "faildelay": faildelay,
             "kwcheck": True,
-            "retry_timeout": nsleep * (retry_period + overhead),
-            "retry_period": retry_period,
+            "retry_timeout": sufficient_timeout,
         }
         with open(self.ctr_file, mode="w") as f:
             f.write("0")
         self.assertTrue(_wsubmain(self.ctr_file, **kw))
 
         kw = {
-            "nsleep": nsleep,
+            "ncausefailure": ncausefailure,
+            "faildelay": faildelay,
             "kwcheck": True,
-            "retry_timeout": nsleep * (retry_period - 0.001),
-            "retry_period": retry_period,
+            "retry_timeout": insufficient_timeout,
         }
         with open(self.ctr_file, mode="w") as f:
             f.write("0")

